### ISSUE TITLE
Installing airflow with extra packages

### ISSUE THREAD

**Issue by JK87iab** [bitnami/charts#2588]
**Which chart**:
airflow

What is the best practice to install additional airflow and python packages? (https://airflow.apache.org/docs/stable/installation.html)

Is the only approach to rebuild the images?

**Comment by javsalgar** [comment]
Hi,

We added this to the container so you could add any other python packages

https://github.com/bitnami/bitnami-docker-airflow/blob/master/1/debian-10/rootfs/run.sh#L14

If you mount a requirements.txt `pip` will be executed. Would that work for your case?

**Comment by JK87iab** [comment]
Thx that worked

**Comment by javsalgar** [comment]
Good to know! Do not hesitate to open a new ticket if you find more issues :D

**Comment by emerson131** [comment]
Ive been having the same issue and i did see the comments on this closed ticket but i wasnt able to successfully mount the requirements.txt file to the location /bitnami/python/requirements.txt since there is an existing python executable at the same location. 

sudo docker run -d --name airflow -p 8080:8080 \
  -e AIRFLOW_FERNET_KEY=xxxxx \
  -e AIRFLOW_EXECUTOR=CeleryExecutor \
  -e AIRFLOW_DATABASE_NAME=xxxx \
  -e AIRFLOW_DATABASE_USERNAME=xxxx \
  -e AIRFLOW_DATABASE_PASSWORD=xxxxx \
  -e AIRFLOW_LOAD_EXAMPLES=no \
  -e AIRFLOW_PASSWORD=xxxx \
  -e AIRFLOW_USERNAME=user \
  -e AIRFLOW_EMAIL=user@example.com \
  --net airflow-tier \
  --volume /home/ec2-user/airflow-persistence:/bitnami \
  --mount type=bind,source=/home/ec2-user/dag,target=/opt/bitnami/airflow/dags,bind-propagation=shared \
  --mount type=bind,source=/home/ec2-user/python,target=/bitnami/python,bind-propagation=shared \
  bitnami/airflow:latest

The python source folder contains the requirements.txt file.

**Comment by ashishmgofficial** [comment]
@javsalgar The requirements.txt is working great in scheduler but the in worker pods im getting module not found error ? Anything im missing ?

**Comment by javsalgar** [comment]
Hi,

This is weird, as I see the same code in the worker container.

https://github.com/bitnami/bitnami-docker-airflow-worker/blob/master/2/debian-10/rootfs/app-entrypoint.sh#L17

Are you installing the exact same packages in one and the other?

**Comment by ashishmgofficial** [comment]
@javsalgar  I have a PV created for holding the requirements.txt. so all the worker and scheduler takes the same requirements.txt

**Comment by javsalgar** [comment]
Hi,

Could you also check if the `pip install` command is being run in both nodes? If it says module not found error it could be related to the module not being installed in the proper folder (which I don't think it's the case) or maybe something incorrect with the `PYTHONPATH` variable. At which point does it fail? Which is the exact traceback?

**Comment by elvin-han** [comment]
Hi, I have same problem.

The requirements.txt is working great in scheduler and web. But not working in worker.

-------------
web:
  baseUrl: http://localhost:8080

  extraVolumeMounts:
    - name: requirements
      mountPath: /bitnami/python/

  extraVolumes:
    - name: requirements
      configMap:
        name: requirements

scheduler:
  extraVolumeMounts:
    - name: requirements
      mountPath: /bitnami/python/

  extraVolumes:
    - name: requirements
      configMap:
        name: requirements

worker:
  extraVolumeMounts:
    - name: requirements
      mountPath: /bitnami/python/

  extraVolumes:
    - name: requirements
      configMap:
        name: requirements
-------------

**Comment by elvin-han** [comment]
<img width="742" alt="스크린샷 2021-03-29 오후 7 33 10" src="https://user-images.githubusercontent.com/20236034/112824581-a9ffc200-90c5-11eb-9674-579faba4500d.png">

**Comment by javsalgar** [comment]
Pinging @miguelaeh so he takes this into account when working on the latest chart changes

**Comment by amal209** [comment]
Hiii, i created a volume for requirements.txt and upgraded the chart, but i'm still getting ModuleNotFoundError: No module named 'yfinance'

**Comment by amal209** [comment]
here is my requirements and values
![image](https://user-images.githubusercontent.com/57093156/162933405-7ef5ab9c-f1e7-422d-9b2f-22caa1256380.png)

**Comment by amal209** [comment]
UPDATE !!
I solved the problem by adding extraVolumeMounts and extraVolumes to the airflow web and scheduler pods

**Comment by amal209** [comment]
Hello,
here's my requirements.txt
[image: image.png]
And i added this in my values.yaml
[image: image.png]

**Comment by jamesmedice** [comment]
> adding extraVolumeMounts and extraVolumes to the airflow web and scheduler pods

in my case i have,
      containers: 
        - name: airflow-web
          image: bitnami/airflow:2.2.3-debian-10-r45
          imagePullPolicy: "IfNotPresent"

          env: 
          envFrom:
          ports:
            - name: http
              containerPort: 8080
          volumeMounts:
            - name: requirements
              mountPath: /bitnami/python/requirements.txt              
              subPath: requirements.txt
          resources:
            limits: {}
            requests: {}
      volumes:
        - name: requirements
          configMap:
            name: airflow-requirements


how would be the extraVolumeMounts and extraVolumes ? should i replace volumes and volumemounts ?

**Comment by vorandrew** [comment]
Why another user is doing `pip install` ?

```
airflow-worker 15:55:42.14 INFO ==>
airflow-worker 15:55:42.14 INFO ==> Welcome to the Bitnami airflow-worker container
airflow-worker 15:55:42.15 INFO ==> Subscribe to project updates by watching https://github.com/bitnami/containers
airflow-worker 15:55:42.15 INFO ==> Submit issues and feature requests at https://github.com/bitnami/containers/issues
airflow-worker 15:55:42.15 INFO ==> Upgrade to Tanzu Application Catalog for production environments to access custom-configured and pre-packaged software components. Gain enhanced features, including Software Bill of Materials (SBOM), CVE scan result reports, and VEX documents. To learn more, visit https://bitnami.com/enterprise
airflow-worker 15:55:42.15 INFO ==>
airflow-worker 15:55:42.16 INFO ==> Enabling non-root system user with nss_wrapper
WARNING: The directory '/opt/bitnami/airflow/.cache/pip' or its parent directory is not owned or is not writable by the current user. The cache has been disabled. Check the permissions and owner of that directory. If executing pip with sudo, you should use sudo's -H flag.
Collecting corsound-airflow@ https://_json_key_base64:****@us-python.pkg.dev/mvp-2023-10-10/corsound/corsound-airflow/corsound_airflow-2.0.0-py3-none-any.whl (from -r /bitnami/python/requirements.txt (line 2))
Downloading https://_json_key_base64:****@us-python.pkg.dev/mvp-2023-10-10/corsound/corsound-airflow/corsound_airflow-2.0.0-py3-none-any.whl (4.5 kB)
Requirement already satisfied: flask in /opt/bitnami/airflow/venv/lib/python3.11/site-packages (from -r /bitnami/python/requirements.txt (line 1)) (2.2.5)
Requirement already satisfied: Werkzeug>=2.2.2 in /opt/bitnami/airflow/venv/lib/python3.11/site-packages (from flask->-r /bitnami/python/requirements.txt (line 1)) (2.2.3)
Requirement already satisfied: Jinja2>=3.0 in /opt/bitnami/airflow/venv/lib/python3.11/site-packages (from flask->-r /bitnami/python/requirements.txt (line 1)) (3.1.4)
Requirement already satisfied: itsdangerous>=2.0 in /opt/bitnami/airflow/venv/lib/python3.11/site-packages (from flask->-r /bitnami/python/requirements.txt (line 1)) (2.2.0)
Requirement already satisfied: click>=8.0 in /opt/bitnami/airflow/venv/lib/python3.11/site-packages (from flask->-r /bitnami/python/requirements.txt (line 1)) (8.1.7)
Collecting peppercorn (from corsound-airflow@ https://_json_key_base64:ewogICJjbGllbnRfZW1haWwiOiAiY29yc291bmQtYXJ0aWZhY3RvcnlAbXZwLTIwMjMtMTAtMTAuaWFtLmdzZXJ2aWNlYWNjb3VudC5jb20iLAogICJjbGllbnRfaWQiOiAiMTEwNTIwMjk4NTQxOTU1MTk5MDIzIiwKICAicHJpdmF0ZV9rZXkiOiAiLS0tLS1CRUdJTiBQUklWQVRFIEtFWS0tLS0tXG5NSUlFdndJQkFEQU5CZ2txaGtpRzl3MEJBUUVGQUFTQ0JLa3dnZ1NsQWdFQUFvSUJBUUN2NjRheU1tTU42ZDRSXG5nRk1KVWk3azV4NkNNUjlrS0daZGxQYmo1QzNQbnhPaEt1VWoySDlNSjRsUzVkSWlWWUp1a1BhV0JsUGF6anA1XG5KQW9oVGRMaVBaa2RGNEJhL3JiVUdVc00wWHpuT1lUOVR4NCtzdVNxVUtKRVpUYjNHTDAwNG1LQUVvcTZJcDd6XG5jVjJ1cStWLzdYRDlnQTZxWmlsdHZGQlFJNnRTZVhCUXBYbEovVnJFdGRIMGc5VCtlU2FiVVBtdVVXNUNlWWtuXG5GRkFnRzBhSWRoS1g5bmdKMmV3R1gzT1oxWXFQNjdBOXNsRlA5RS9ERSt0eXFJcTJYcGVjdWZEdDJHNjhhbENtXG5vN1JyQTZJS09BRGRuZVNySVlVVUJlbSsrdzRiNmwzMUxHVDN3dllpWjBxU1VVaVZzNy9sSE1qRkQvbjJOajhmXG5iamoyTjd0ZEFnTUJBQUVDZ2dFQVVPWE0wNzVUQ2M5dHBQdXZjNzVjdlR2QkxscWJsbW56cFA0L1NNNGgrUE9zXG55YlVabVZIU1BEbUVMMkJyTVFuRWlNYlY5bWlqZ0IvTkNPODNUMWhqN1ZFcFM1VUduQkc5RHEwdzg0eVVWUnFBXG4yeUhhaFRlcjZ0eFl1bjRXK0N0aXV1S09MRm9hZkgwMlBqbjRQRDVsTTk2RFVyZmtScHpqV1VpN1lCYi9EbFh3XG5vZXFZWjVaQ0Qyd0Q3dDV1NFV0YUgzcW5hamVRUElzV2E0TzlPdXVwQW1DODhMWTh6bDhlL2VrZ2REcHJiSmxXXG5FUWo1dm1wZjRiNzZKbTNQR0FQMU4vWWlTSlZmejFldG5CVFZyR1dhazZZTEU4Sk9ZeHptSU51aWh2dXhUcHpjXG5XSkdEd0lWSmxWRXpNdjhiM09lQ1hCSjFyWmpKTThRMzNZeEdTeVEyQ3dLQmdRRHMzQklsL29jdlVvVFJSVS9vXG5vMjgyMXhQQTJVWGtRZUdTaDI5VWc1Nms5WjM4QU5GaG9ONDJUSDVuM0ptNjZBVDRyTVFMdEJCMkRKY0djeXppXG5vQUpDMVJzSEZDcEVnK1hxRmxXTnBPNmIrL2pZenRCMEp0VytiZXRiZjMxdHJzZlVKcGEwZFA4NnQ4b2RzMWcwXG5tQWxWQkkvVFJ0OUpNQ1B3VlpIa05wUXdGd0tCZ1FDK0lzc3d1SlJNVUVrT1NPK0ZIZk1KVUoyQnhsUW1UR0xxXG5pUU9YWnNjVnpWc05YYXlydHhwRkJyK0hNWnk4TjExK0s2b1Z3YlREY1psM1lPZlpmZ1dLdVdHU3hibkQwcnV1XG52elVrUzdGRUFNL1JoZFhrK3QzdHQzYk1IYkNNVWx2d2g4VzM1bytLcDBsOUFXZ1o0Ykx2WkxNVmZhamJ0M3FUXG5UbXhQL3liRXF3S0JnUURRRy9SZUthckJhM0dESmJFMktCTVpUaFB2QjNnZk83dGtNVmNpSEpzUk9UVmNLNlBNXG5jYm1NWDkxWWlEMFNPYlhVZ0llbitNZlYwTzRKSHIwN1NYZHpIU2FvSzVHUHJZd2ZCd3lZYldldzZjMFZRWEZwXG4vSTBreWhOV3kwZDdUV3dKL1BYQkRHTU1SbHR4aVJoZzdZeVZoSEtTSWQwaHd3aHdxOTZKYzJMMGRRS0JnUUM5XG45N2Q1dzBGd0Nkc2tLWjBlTERFeUFsaU1XK0toclRvN1NJL05Da1FwTlZjZXdKUVBzTTdQWHBHbit0cTFkcit6XG5tdDdDQkJxZUYzY1V2ajZIb0JEYmJLTGZhU3hVU2MxNUo5TnBvekxHVmVWTld1RzllU1NaS08vSnVuUTFqdWdMXG5KeXhWdEVlQWFlblR1cjhPOU9DNXJLdjg4ZUx2TzU5OVpDSjF2dUUySndLQmdRRElVaGVRdG56UUVvcDlwUGtuXG4wbzI1K0JjZ003R2s0SndkOWdZckZwMCttTHhDTmNjWTIrdFFKdVFLUWd4UDU3YzNhT3Z0TTdOaXVBSDViVUhHXG5wYU0vQk9JU3I5VktpUTJtVTFrZDdsU3JqYWZJQ21jQ1g2MGh2TFJmRUovc1FCR1czNHUwWmZ3VjJ3ZStMMnMrXG44UXZaTDUza3F0YmpaK1UwcTFHdm42OUNqQT09XG4tLS0tLUVORCBQUklWQVRFIEtFWS0tLS0tXG4iLAogICJwcml2YXRlX2tleV9pZCI6ICI0NTY3Y2QxNzIwNDgzYzA3YWM0ODMwODNmZWM3Y2I4NDVjMGIxMzczIiwKICAidG9rZW5fdXJpIjogImh0dHBzOi8vb2F1dGgyLmdvb2dsZWFwaXMuY29tL3Rva2VuIiwKICAidHlwZSI6ICJzZXJ2aWNlX2FjY291bnQiCn0=@us-python.pkg.dev/mvp-2023-10-10/corsound/corsound-airflow/corsound_airflow-2.0.0-py3-none-any.whl->-r /bitnami/python/requirements.txt (line 2))
Downloading peppercorn-0.6-py3-none-any.whl.metadata (3.4 kB)
Requirement already satisfied: MarkupSafe>=2.0 in /opt/bitnami/airflow/venv/lib/python3.11/site-packages (from Jinja2>=3.0->flask->-r /bitnami/python/requirements.txt (line 1)) (2.1.5)
Downloading peppercorn-0.6-py3-none-any.whl (4.8 kB)
Installing collected packages: peppercorn, corsound-airflow
ERROR: Could not install packages due to an OSError: [Errno 30] Read-only file system: '/opt/bitnami/airflow/venv/lib/python3.11/site-packages/peppercorn'
[notice] A new release of pip is available: 24.0 -> 24.1.2
[notice] To update, run: pip install --upgrade pip
```

**Comment by vorandrew** [comment]
Why another user is doing `pip install` ?

```
airflow-worker 15:55:42.14 INFO ==>
airflow-worker 15:55:42.14 INFO ==> Welcome to the Bitnami airflow-worker container
airflow-worker 15:55:42.15 INFO ==> Subscribe to project updates by watching https://github.com/bitnami/containers
airflow-worker 15:55:42.15 INFO ==> Submit issues and feature requests at https://github.com/bitnami/containers/issues
airflow-worker 15:55:42.15 INFO ==> Upgrade to Tanzu Application Catalog for production environments to access custom-configured and pre-packaged software components. Gain enhanced features, including Software Bill of Materials (SBOM), CVE scan result reports, and VEX documents. To learn more, visit https://bitnami.com/enterprise
airflow-worker 15:55:42.15 INFO ==>
airflow-worker 15:55:42.16 INFO ==> Enabling non-root system user with nss_wrapper
WARNING: The directory '/opt/bitnami/airflow/.cache/pip' or its parent directory is not owned or is not writable by the current user. The cache has been disabled. Check the permissions and owner of that directory. If executing pip with sudo, you should use sudo's -H flag.
Collecting corsound-airflow@ https://_json_key_base64:****@us-python.pkg.dev/mvp-2023-10-10/corsound/corsound-airflow/corsound_airflow-2.0.0-py3-none-any.whl (from -r /bitnami/python/requirements.txt (line 2))
Downloading https://_json_key_base64:****@us-python.pkg.dev/mvp-2023-10-10/corsound/corsound-airflow/corsound_airflow-2.0.0-py3-none-any.whl (4.5 kB)
Requirement already satisfied: flask in /opt/bitnami/airflow/venv/lib/python3.11/site-packages (from -r /bitnami/python/requirements.txt (line 1)) (2.2.5)
Requirement already satisfied: Werkzeug>=2.2.2 in /opt/bitnami/airflow/venv/lib/python3.11/site-packages (from flask->-r /bitnami/python/requirements.txt (line 1)) (2.2.3)
Requirement already satisfied: Jinja2>=3.0 in /opt/bitnami/airflow/venv/lib/python3.11/site-packages (from flask->-r /bitnami/python/requirements.txt (line 1)) (3.1.4)
Requirement already satisfied: itsdangerous>=2.0 in /opt/bitnami/airflow/venv/lib/python3.11/site-packages (from flask->-r /bitnami/python/requirements.txt (line 1)) (2.2.0)
Requirement already satisfied: click>=8.0 in /opt/bitnami/airflow/venv/lib/python3.11/site-packages (from flask->-r /bitnami/python/requirements.txt (line 1)) (8.1.7)
Collecting peppercorn (from corsound-airflow@ https://_json_key_base64:=@us-python.pkg.dev/mvp-2023-10-10/corsound/corsound-airflow/corsound_airflow-2.0.0-py3-none-any.whl->-r /bitnami/python/requirements.txt (line 2))
Downloading peppercorn-0.6-py3-none-any.whl.metadata (3.4 kB)
Requirement already satisfied: MarkupSafe>=2.0 in /opt/bitnami/airflow/venv/lib/python3.11/site-packages (from Jinja2>=3.0->flask->-r /bitnami/python/requirements.txt (line 1)) (2.1.5)
Downloading peppercorn-0.6-py3-none-any.whl (4.8 kB)
Installing collected packages: peppercorn, corsound-airflow
ERROR: Could not install packages due to an OSError: [Errno 30] Read-only file system: '/opt/bitnami/airflow/venv/lib/python3.11/site-packages/peppercorn'
[notice] A new release of pip is available: 24.0 -> 24.1.2
[notice] To update, run: pip install --upgrade pip
```
